src/server/worker/worker.go

```diff
@@ -1,29 +1,198 @@
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/auth"
 	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/backoff"
+	"github.com/pachyderm/pachyderm/src/server/pkg/dlock"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"github.com/pachyderm/pachyderm/src/server/worker/driver"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+	"github.com/pachyderm/pachyderm/src/server/worker/pipeline/service"
+	"github.com/pachyderm/pachyderm/src/server/worker/pipeline/spout"
+	"github.com/pachyderm/pachyderm/src/server/worker/pipeline/transform"
+	"github.com/pachyderm/pachyderm/src/server/worker/server"
+	"github.com/pachyderm/pachyderm/src/server/worker/stats"
 )
 
-// MatchDatum checks if a datum matches a filter.  To match each string in
-// filter must correspond match at least 1 datum's Path or Hash. Order of
-// filter and data is irrelevant.
-func MatchDatum(filter []string, data []*pps.InputFile) bool {
-	// All paths in request.DataFilters must appear somewhere in the log
-	// line's inputs, or it's filtered
-	matchesData := true
-dataFilters:
-	for _, dataFilter := range filter {
-		for _, datum := range data {
-			if dataFilter == datum.Path ||
-				dataFilter == base64.StdEncoding.EncodeToString(datum.Hash) ||
-				dataFilter == hex.EncodeToString(datum.Hash) {
-				continue dataFilters // Found, move to next filter
+const (
+	masterLockPath = "_master_worker_lock"
+)
+
+// The Worker object represents
+type Worker struct {
+	APIServer *server.APIServer // Provides rpcs for other nodes in the cluster
+	driver    driver.Driver     // Provides common functions used by worker code
+	namespace string            // The namespace in which pachyderm is deployed - TODO: what is this for?
+	status    *transform.Status // An interface for inspecting and canceling the actively running task
+}
+
+// NewWorker constructs a Worker object that provides all worker functionality:
+//  1. a master goroutine that attempts to obtain the master lock for the pipeline workers and direct jobs
+//  2. a worker goroutine that gets tasks from the master and processes them
+//  3. an api server that serves requests for status or cross-worker communication
+//  4. a driver that provides common functionality between the above components
+func NewWorker(
+	pachClient *client.APIClient,
+	etcdClient *etcd.Client,
+	etcdPrefix string,
+	pipelineInfo *pps.PipelineInfo,
+	workerName string,
+	namespace string,
+	hashtreePath string,
+) (*Worker, error) {
+	stats.InitPrometheus()
+
+	hasDocker := true
+	if _, err := os.Stat("/var/run/docker.sock"); err != nil {
+		hasDocker = false
+	}
+
+	driver, err := driver.NewDriver(
+		pipelineInfo,
+		pachClient,
+		etcdClient,
+		etcdPrefix,
+		filepath.Join(hashtreePath, uuid.NewWithoutDashes()),
```

Comment:
We should think about if we want the server to use the same cache when it gets restarted or if we want to do some sort of cleanup of older cache directories on startup.